pkg/util/kms.go

```diff
@@ -0,0 +1,277 @@
+	for k, v := range config {
+		vaultConfig[k] = v
+	}
+
+	// create TLS files out of secrets
+	// replace in vaultConfig secret names with local temp files paths
+	err := tlsConfig(vaultConfig, namespace)
+	if err != nil {
+		return nil, fmt.Errorf(`❌ Could not init vault tls config %q in namespace %q`, config, namespace)
+	}
+
+	// veify backend path, use default value if not set
+	if b, ok := config[VaultBackendPath]; !ok || b == "" {
+		log.Infof("KMS: using default backend path %v", defaultVaultBackendPath)
+		vaultConfig[VaultBackendPath] = defaultVaultBackendPath
+	}
+
+	// fetch vault token from the secret
+	secret := KubeObject(bundle.File_deploy_internal_secret_empty_yaml).(*corev1.Secret)
+	secret.Namespace = namespace
+	secret.Name = tokenSecretName
+	if !KubeCheck(secret) {
+		return nil, fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+	}
+	token := secret.StringData["token"]
+	vaultConfig[vaultToken] = token
+
+	return vault.New(vaultConfig)
+}
+
+// tlsConfig create temp files with TLS keys and certs
+func tlsConfig(config map[string]interface{}, namespace string) (error) {
+	secret := KubeObject(bundle.File_deploy_internal_secret_empty_yaml).(*corev1.Secret)
+	secret.Namespace = namespace
+
+	if caCertSecretName, ok := config[vaultCaCert]; ok {
+		secret.Name = caCertSecretName.(string)
+		if !KubeCheckOptional(secret) {
+			return fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+		}
+		caFileAddr, err := writeCrtsToFile(secret.Name, namespace, secret.Data["cert"], vaultCaCert)
+		if err != nil {
+			return fmt.Errorf("can not write crt %v to file %v", vaultCaCert, err)
+		}
+		config[vaultCaCert] = caFileAddr
+	}
+
+	if clientCertSecretName, ok := config[vaultClientCert]; ok {
+		secret.Name = clientCertSecretName.(string)
+		if !KubeCheckOptional(secret) {
+			return fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+		}
+		clientCertFileAddr, err := writeCrtsToFile(secret.Name, namespace, secret.Data["cert"], vaultClientCert)
+		if err != nil {
+			return fmt.Errorf("can not write crt %v to file %v", vaultClientCert, err)
+		}
+		config[vaultClientCert] = clientCertFileAddr
+
+	}
+	if clientKeySecretName, ok := config[vaultClientKey]; ok {
+		secret.Name = clientKeySecretName.(string)
+		if !KubeCheckOptional(secret) {
+			return fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+		}
+		clientKeyFileAddr, err := writeCrtsToFile(secret.Name, namespace, secret.Data["key"], vaultClientKey)
+		if err != nil {
+			return fmt.Errorf("can not write crt %v to file %v", vaultClientKey, err)
+		}
+		config[vaultClientKey] = clientKeyFileAddr
+	}
+	return nil
+}
+
+// PutSecret writes the secret to the secrets store
+func PutSecret(client secrets.Secrets, secretName, secretValue, secretPath string) error {
+
+	keyContext := map[string]string{}
```

Comment:
what is keyContext? why is it empty? for both putSecret and GetSecret functions

Reply:
Special key access options, so far not needed 😊.

Reply:
Used for special key access options, so far not needed 😊.

Reply:
can you attach a link / relevant doc?

Reply:
Yep, we could pass nil as a context for now. See [commit](https://github.com/noobaa/noobaa-operator/pull/797/commits/34b5de67314ed22a0b645bcae7e90f4736816be1).
Here is an example of a special key:
https://github.com/libopenstorage/secrets/commit/dde442ea20ec9d59c71cea5ee0f21eeffd17ed19